File: Frontend/Javascript/neetcode/hashing/3.Two Sum.js
// Brute Force
function twoSum(nums, target) {
  for (let i = 0; i < nums.length; i++) {
    for (let j = i + 1; j < nums.length; j++) {
      if (nums[i] + nums[j] == target) return [i, j];
    }
  }
  return [];
}

//HashMap one pass
function twoSum2(nums, target) {
  const map = new Map();
  for (let i = 0; i < nums.length; i++) {
    const diff = target - nums[i];
    if (map.has(diff)) return [i, map.get(diff)];
    else map.set(nums[i], i);
  }
  return [];
}

console.log(twoSum((nums = [3, 4, 5, 6]), (target = 7)));
console.log(twoSum2((nums = [3, 4, 5, 6]), (target = 7)));

File: C#/NeetCode/1. Hashing/3_Two_Sum.cs
namespace NeetCode._1._Hashing
{
    public class _3_Two_Sum
    {
        public _3_Two_Sum()
        {
            
        }

        public int[] BruteForce(int[] nums, int target)
        {
            for (int i = 0; i < nums.Length; i++)
            {
                for (int j = i + 1; j < nums.Length; j++)
                {
                    if (nums[i] + nums[j] == target)
                    {
                        return new int[] { i, j };
                    }
                }
            }
            return new int[0];
        }

        public int[] Sorting(int[] nums, int target)
        {
            List<int[]> bag = new List<int[]>();
            for(int i = 0; i < nums.Length; i++)
            {
                bag.Add(new int[] { nums[i], i });
            }

            bag.Sort((a, b) => a[0].CompareTo(b[0]));

            int a = 0; int b = nums.Length - 1;
            while(a < b)
            {
                int diff = bag[b][0] - bag[a][0];
                if (diff > target)
                    b--;
                else if(diff < target) 
                    a++;
                else
                    return new int[] { Math.Min(bag[a][1], bag[b][1]), Math.Max(bag[a][1], bag[b][1]) };
            }

            return new int[0];
        }
    }
}
